overlord/README.md

```diff
@@ -0,0 +1,80 @@
+
+ The goals of the `state.Change` mechanisms is that operations should survive restarts and reboots and that on error snapd should try to bring back the external state to a previous good state if possible.
+
+`state.TaskRunner`
+-------------------
+The `state.TaskRunner` is responsible for `state.Change` and `state.Task` execution, and their state management. The do and undo logic of `state.Task`s is defined by `Task` kind using `TaskRunner.AddHandler`.
+
+ During execution, a `Task` goes through a series of statuses. These are represented by state.Status and will finish in a ready status of either `DoneStatus, UndoneStatus, ErrorStatus` or `HoldStatus`.
+
+ If errors are encountered, the `TaskRunner `will normally try to recursively execute the undo logic of any previously depended-upon `Task`s with the exception of the `Task` that generated the error. It is instead expected that any desired undo should be part of its error paths.
+
+Different `Change`s and independent `Task`s are normally executed concurrently.
+
+`Task`s and `State` locking and consistency
+--------------------------------------------
+Currently, the `Task`  do and undo handlers are started without holding the State lock, but to simplify consistency, it's easier if a `Task` executes while holding the State lock.
+Strictly, `State` lock must only be released when performing slow operations, such as:
+-   copying, compressing or uncompressing of large amounts of on-disk data
+-   network operations
+
+So in practice, most handler code should start with:
+
+```
+st.Lock()
+defer st.Unlock()
+```
+
+where `st` is the runtime `state.State` instance, accessible via `Task.State() or the handler manager.
+
+The deferred `Unlock` will implicitly commit any working state mutations at the end of the handler.
+
+Due to potential restarts, the do or undo handler logic in a Task may be re-executed if it hasn't already completed. This necessitates the following considerations:
+-   on-disk/external state manipulation should be idempotent or equivalent
+-   working state manipulation should either be idempotent or designed to combine working state mutations with setting the next status of the task. This approach currently requires using `Task.SetStatus` before returning form the handler
+
+If slow operations need to be performed, the required `Unlock/Lock` should happen before any working state manipulation.
+
+If the `State` lock is released during a handler, the code needs to consider that other code could have manipulated some relevant working state.There may be also cases where it’s neither possible nor desirable to hold the `State` lock for the entirety of a state manipulation, such as when a manipulation spans multiple subsystems, and so spans multiple tasks. For all such cases, and to simplify reasoning, snapd offers other coordination mechanisms with differing granularity to the `State` lock.
```

Comment:
Sounds slightly better like this to me but I'm nitpicking

```suggestion
If the `State` lock is released in a handler, the code needs to consider that other code may have manipulated some relevant working state. There may be also cases where it’s neither possible nor desirable to hold the `State` lock for the entirety of a state manipulation, such as when a manipulation spans multiple subsystems, and so spans multiple tasks. For all such cases, and to simplify reasoning, snapd offers other coordination mechanisms with differing granularity to the `State` lock.
```

Reply:
I added a released and reacquired, the "could" was formulated thinking about the point of reacquired the lock